test/langtools/tools/javac/util/IteratorsTest.java

```diff
@@ -0,0 +1,172 @@
+        TestIterator<String> test1 = new TestIterator<>(List.of("1").iterator());
+        TestIterator<String> test2 = new TestIterator<>(List.of("2").iterator());
+        Iterator<String> compound = Iterators.createCompoundIterator(List.of(test1, test2), c);
+
+        //nothing should be called before the hasNext or next is called:
+        assertAndResetMaxCalls(c, 0);
+        assertAndResetMaxCalls(test1, 0, 0);
+        assertAndResetMaxCalls(test2, 0, 0);
+
+        //when hasNext is called, should invoke the hasNext delegate once:
+        Assertions.assertTrue(compound.hasNext());
+
+        assertAndResetMaxCalls(c, 1);
+        assertAndResetMaxCalls(test1, 1, 0);
+        assertAndResetMaxCalls(test2, 0, 0);
+
+        Assertions.assertTrue(compound.hasNext());
+
+        assertAndResetMaxCalls(c, 0);
+        assertAndResetMaxCalls(test1, 1, 0);
+        assertAndResetMaxCalls(test2, 0, 0);
+
+        //next may invoke hasNext once:
+        Assertions.assertEquals("1", compound.next());
+
+        assertAndResetMaxCalls(c, 0);
+        assertAndResetMaxCalls(test1, 1, 1);
+        assertAndResetMaxCalls(test2, 0, 0);
+    }
+
+    private void assertAndResetMaxCalls(TestIterator<?> test, int maxExpectedHasNextCalls, int maxExpectedNextCalls) {
+        if (test.hasNextCalls > maxExpectedHasNextCalls) {
+            Assertions.fail("too many hasNext invocations: " + test.hasNextCalls +
+                            ", expected: " + maxExpectedHasNextCalls);
+        }
+        test.hasNextCalls = 0;
+        if (test.nextCalls > maxExpectedNextCalls) {
+            Assertions.fail("too many next invocations: " + test.nextCalls +
+                            ", expected: " + maxExpectedNextCalls);
+        }
+        test.nextCalls = 0;
+    }
+
+    private void assertAndResetMaxCalls(TestConverter<?, ?> test, int maxExpectedApplyCalls) {
+        if (test.applyCalls > maxExpectedApplyCalls) {
+            Assertions.fail("too many apply invocations: " + test.applyCalls +
+                            ", expected: " + maxExpectedApplyCalls);
+        }
+        test.applyCalls = 0;
+    }
+
+    static class TestIterator<T> implements Iterator<T> {
+        int hasNextCalls;
+        int nextCalls;
+        final Iterator<T> delegate;
+
+        public TestIterator(Iterator<T> delegate) {
+            this.delegate = delegate;
+        }
+
+        @Override
+        public boolean hasNext() {
+            hasNextCalls++;
+            return delegate.hasNext();
+        }
+
+        @Override
+        public T next() {
+            nextCalls++;
+            return delegate.next();
+        }
+    }
+
+    static class TestConverter<I, O> implements Function<I, Iterator<O>> {
+        int applyCalls;
+        final Function<I, Iterator<O>> delegate;
+
+        public TestConverter(Function<I, Iterator<O>> delegate) {
+            this.delegate = delegate;
```

Comment:
This parameter doesn't really seem to be used in the test?

Reply:
I suggest this TestConverter to take a single type parameter `T` and implement `Function<TestIterator<S>, Iterator<S>>`